include/splay.h

```diff
@@ -100,47 +100,132 @@ template<class V>
 void
 SplayNode<V>::walk(SPLAYWALKEE * walkee, void *state)
 {
-    if (left)
-        left->walk(walkee, state);
-
-    walkee(data, state);
-
-    if (right)
-        right->walk(walkee, state);
+    // in-order walk through tree using Morris Traversal
+    SplayNode<V> *cur = const_cast<SplayNode<V> *>(this);
+    while (cur != nullptr) {
+        if (cur->left == nullptr) {
+            // no left subtree, call walkee on cur->data and
+            // descent into right subtree
+            walkee(cur->data, state);
+            cur = cur->right;
+        } else {
+            // first descent into left subtree
+
+            // find right-most child in left tree
+            SplayNode<V> *rmc = cur->left;
+            while (rmc->right && rmc->right != cur)
+                rmc = rmc->right;
+
+            if (rmc->right == nullptr) {
+                // found the right most node in the left subtree:
+                // create a temporay thread back to cur to have
+                // a way back out of the subtree
+                rmc->right = cur;
```

Comment:
IMHO walk and visit are fragile at best (also in the recursive variant) :-(
Every call to splay (insert, remove, find) will rotate parts of the tree and break these functions. I don't think there is a safe way to walk/visit the tree without locking modifications first.
That's also true for the iterator SplayConstIterator. Maybe the comment "XXX: This does not seem to iterate the whole thing in some cases." comes from this.
A back-pointer could be used for a more robust approach but it has to be updated for all involved nodes on every splay-call, too. This would not only mean more memory usage but also more cpu.

Reply:
> Every call to splay (insert, remove, find) will rotate parts of the tree and break these functions

Well, we are not supposed to modify a being-iterated container. That rule is true for nearly all standard containers as well. I do not think we need to worry about this particular problem.

> A back-pointer could be used for a more robust approach but it has to be updated for all involved nodes on every splay-call, too.

I was thinking that it does not need to be updated. It must be nil everywhere except inside iterators. In a sense, that pointer is not a part of the tree. It is a part of the iterator.

Reply:
To avoid possible performance issues, I've followed your proposal and updated `visit()` to use a private pointer `visit_thread_up` instead of the right child pointer. As a side-effect `visit()` now can distinguish between down and up movements and therefor avoid the second search for the right most child.
Unfortunately we still can't use `visit()` (or a variant of it working on nodes) to destroy the tree, because the visitor would have to call `delete` on the root node (`this`) in-order before it descents into the right subtree which would be fatal.
On my test system with a sorted list of 8 million ip addresses the memory usage of the squid parent process increased from 1116 MB to 1238 MB and the memory of the single kid from 1133 MB to 1254 MB.

While waiting for the removal of `walk()` I implemented it using `visit()` to reduce the duplicate code.
I've also removed the unused SplayIterator class that's missing an implementation.